rpyc/lib/win_unix_socket.py

```diff
@@ -0,0 +1,540 @@
+
+def wait_for_fd(s, r_w_connect, timeout):
+    """
+    Wait until the socket is ready.
+
+    Low on error handling.
+
+    r_w_connect is 0 for read, 1 for write, 2 for connect
+    """
+    p = pollfd()
+    p.fd = s._sock_fd
+    if r_w_connect == 0:
+        p.events = POLLIN
+    elif r_w_connect == 1:
+        p.events = POLLOUT
+    else:
+        p.events = POLLIN | POLLERR
+
+    return poll(ctypes.byref(p), 1, int(timeout))
+
+
+def call_socket_fn(s, r_w_connect, sockfn):
+    """
+    Call a function (passed in as a lambda: () -> bool) on a socket,
+    with a specified timeout in ms.
+    r_w_connect is 0 for read, 1 for write, 2 for connect
+    The socket must be non-blocking unless timeout is -1.
+    sockfn should be a lambda that returns non-zero on success and zero on fail.
+    Heavily based on cpython's sock_call_ex.
+    """
+    # QueryPerformanceCounter, monotonic and accurate to sub-microsecond.
+    if s._timeout is None:
+        timeout = -1
+    else:
+        timeout = s._timeout
+    if timeout >= 0:
+        end_t = time.clock() + timeout
+
+    while True:
+        # wait until we can do the thing; always wait if we're connecting
+        if timeout > 0 or r_w_connect == 2:
+            if timeout < 0:
+                milliseconds = -1
+            else:
+                milliseconds = max(1, int((end_t - time.clock()) * 1000.0))
+            err = wait_for_fd(s, r_w_connect, milliseconds)
+            if err < 0:
+                # get the error reason
+                err = errno_fn()
+                if err == WSAEINTR:
+                    # interrupted by signal (and didn't raise) => retry
+                    continue
+                else:
+                    # some other kind of error => raise
+                    raise_socket(err)
+            elif err == 0:
+                # timeout
+                raise socket.timeout("timed out")
+
+        # socket is ready; do the thing (loop until we aren't interrupted)
+        # the loop sets 'err' if we break out of the loop.
+        while True:
+            ok = (sockfn() != 0)
+            if ok:
+                ######################
+                # The thing is done!
+                # This is the only success codepath.
+                ######################
+                return
+            else:
+                # why did it fail?
+                err = errno_fn()
+                if err == WSAEINTR:
+                    # interrupted by signal (and didn't raise) => retry
+                    continue
+                elif timeout > 0 and err == WSAEWOULDBLOCK:
+                    # break this loop, continue outer loop that does a poll()
+                    # and waits for the async IO to finish.
+                    break
+                else:
+                    # if we're here, the error is scary
+                    raise_socket(err)
+
+
+def _make_address(path: bytes):
```

Comment:
does this parse in py 2.7?

Reply:
It doesn't. I fixed this as you were reviewing